### PR TITLE
Task/14

### PR DESCRIPTION
**Pull request by carmopi**

### CODE REVIEW

File: src/main/java/com/carmen/app/boundary/CarResource.java [hunk 1]
```diff
@@ -46,11 +49,34 @@
...
 	@GET
-	public Response getCars() {
-		List<CarDto> cars = this.carService.getCars().stream().map(car -> car.convertToDto()).collect(Collectors.toList());
-		Response response = Response.status(Status.OK).entity(cars).build();
+	public Response getCars(@DefaultValue("1") @QueryParam(value = "page") int page,
+			@DefaultValue("5") @QueryParam(value = "size") int size,
+			@DefaultValue("") @QueryParam(value = "filterBy") String filterBy,
+			@QueryParam(value = "orderBy") String orderBy) {
+		Response response = null;
+		if (page < 1 || size < 0) {
```
Comment (rafapaezbas) [inline src/main/java/com/carmen/app/boundary/CarResource.java:57]: What happens if size is 0?

File: src/main/java/com/carmen/app/boundary/ICarResource.java [hunk 2]
```diff
@@ -20,12 +20,16 @@
...
 	 * 
 	 * Get a list of cars existing in the database
+	 * @param page number of the pagination
+	 * @param size number of cars
+	 * @param filter field to filter
+	 * @param sort  field to order
 	 * 
 	 * @return response that contains a list with all cars values
 	 */
 	@Operation(summary = "Get all cars", responses = {
```
Comment (rafapaezbas) [inline src/main/java/com/carmen/app/boundary/ICarResource.java:30]: It doesnt really get all cars anymore :)

File: src/main/java/com/carmen/app/boundary/CarResource.java [hunk 3]
```diff
@@ -1,20 +1,22 @@
 package com.carmen.app.boundary;
 
-import java.util.ArrayList;
+
+import java.util.LinkedHashMap;
```
Comment (rafapaezbas) [inline src/main/java/com/carmen/app/boundary/CarResource.java:4]: Could be a hashmap aswell instead of a LinkedHaspmap? That is an interesting question I think -> https://stackoverflow.com/questions/2889777/difference-between-hashmap-linkedhashmap-and-treemap

File: src/main/java/com/carmen/app/control/PersistenceService.java [hunk 4]
```diff
@@ -1,54 +1,79 @@
...
-	public List<T> getAll(String NameQuery, Class<T> c){
-		TypedQuery<T> query = this.em.createNamedQuery(NameQuery, c);
-		return query.getResultList();
+
+	public TypedQuery<T> getAll(Class<T> c, Map<String, String> filterMap, String orderBy) {
+		CriteriaBuilder cb = em.getCriteriaBuilder();
+		CriteriaQuery<T> cq = cb.createQuery(c);
+		Root<T> r = cq.from(c);
+		cq.select(r);
+		List<Predicate> predicates = new ArrayList<Predicate>();
```
Comment (rafapaezbas) [inline src/main/java/com/carmen/app/control/PersistenceService.java:41]: Usually namedQueries are more clear in terms of code readability. The are much closer to SQL.

File: src/test/java/com/carmen/app/control/CarServiceTest.java [hunk 5]
```diff
@@ -52,16 +55,16 @@ public void setUp() throws Exception {
...
 	@Test
 	public void testCarCreated() {
-		doNothing().when(this.persistenceService).createNew(car);
+		when(this.persistenceService.createNew(car)).thenReturn(car);
 		assertEquals(car, carService.createCar(car));
 		verify(this.persistenceService).createNew(car);
 	}
 
 	@Test
 	public void testGetCars() {
```
Comment (rafapaezbas) [inline src/test/java/com/carmen/app/control/CarServiceTest.java:64]: Its a good idea to test corner cases of your method. Usually values like NULL, 0, negative numbers or very large numbers are corner cases to check, and unit tests is a nice way to do it.

File: src/test/java/com/carmen/app/control/CarServiceTest.java [hunk 6]
```diff
@@ -52,16 +55,16 @@ public void setUp() throws Exception {
...
-		when(this.persistenceService.getAll("Car.FindCars", Car.class)).thenReturn(new ArrayList<Car>());
+		when(this.persistenceService.getAll(Mockito.eq(Car.class),Mockito.anyMap(),Mockito.anyString())).thenReturn(query);
 		List<Car> expectedCars = new ArrayList<Car>();
-		List<Car> cars = this.carService.getCars();
+		List<Car> cars = this.carService.getCars(1 , 2, "Seat", "brand");
 		assertEquals(expectedCars, cars);
 
 	}
```
Comment (rafapaezbas) [inline src/test/java/com/carmen/app/control/CarServiceTest.java:70]: In general great work! One tip for your code: try to make methods as shorter, method with more that 10 lines are more difficult to read, I like to recommend Clean Code by Robert C. Martin is an excellent read on the topic, specially chapter 3 on functions I find it very interesting and can give you an interesting point of view.

Keep up the good work!
